### PR TITLE
Multiclass Semantic Segmentation using DeepLabV3+

### PR DESCRIPTION
**Pull request by soumik12345**
Semantic segmentation with the goal to assign semantic labels to every pixel in an image is one of the fundamental topics in computer vision. Deep convolutional neural networks based on the Fully Convolutional Neural Network show striking improvement over systems relying on hand-crafted features on benchmark tasks. In this example, we would implement
the **DeepLabV3+** model as proposed by the paper [Encoder-Decoder with Atrous Separable Convolution for Semantic Image Segmentation](https://arxiv.org/pdf/1802.02611.pdf) for multi-class semantic segmentation.

Please find Colab Notebook [here](https://colab.research.google.com/drive/1D_chASKoLqfneEzWQYGRbFLN-yvOyjiG?usp=sharing)

### CODE REVIEW

**Comment by soumik12345** [comment]
Hi @fchollet can you please review this example?

File: examples/vision/deeplabv3_plus.py [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+Title: Multiclass Semantic Segmentation using DeepLabV3+
+Author: [Soumik Rakshit](http://github.com/soumik12345)
+Date created: 2021/08/31
+Last modified: 2021/09/1
+Description: Implement DeepLabV3+ architecture for Multi-class Semantic Segmentation
+"""
+"""
+## Introduction
+"""
+
```
Comment (fchollet) [inline examples/vision/deeplabv3_plus.py:11]: You can remove the break between the two text blocks (it's just one block)

File: examples/vision/deeplabv3_plus.py [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+"""
+Semantic segmentation with the goal to assign semantic labels to every pixel in an image
+is one of the fundamental topics in computer vision. Deep convolutional neural networks
+based on the Fully Convolutional Neural Network show striking improvement over systems
+relying on hand-crafted features on benchmark tasks. In this example, we would implement
+the **DeepLabV3+** model as proposed by the paper [Encoder-Decoder with Atrous Separable
```
Comment (fchollet) [inline examples/vision/deeplabv3_plus.py:17]: Make sure to keep markdown links on a single line, and generally follow [these guidelines](https://github.com/keras-team/keras-io/blob/master/contributor_guide.md)

File: examples/vision/deeplabv3_plus.py [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+import cv2
+import numpy as np
+from glob import glob
+import tensorflow as tf
+from scipy.io import loadmat
+import matplotlib.pyplot as plt
+
+"""shell
+!gdown https://drive.google.com/uc?id=1B9A9UCJYMwTL4oBEo4RZfbMZMaZhKJaz
+!unzip -q instance-level-human-parsing.zip
```
Comment (fchollet) [inline examples/vision/deeplabv3_plus.py:36]: Remove the `!`, it gets added automatically

File: examples/vision/deeplabv3_plus.py [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+train_masks = sorted(glob(os.path.join(DATA_DIR, "Category_ids/*")))[:MAX_IMAGES]
+
+
+def read_image(image_path, mask=False):
+    image = tf.io.read_file(image_path)
+    if mask:
+        image = tf.image.decode_png(image, channels=1)
+        image.set_shape([None, None, 1])
+        image = tf.image.resize(images=image, size=[IMAGE_SIZE, IMAGE_SIZE])
+        image = tf.cast(image, tf.float32)
```
Comment (fchollet) [inline examples/vision/deeplabv3_plus.py:68]: Is the cast necessary? The output of `resize` should be float32

File: examples/vision/deeplabv3_plus.py [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+    image = read_image(image_list)
+    mask = read_image(mask_list, mask=True)
+    return image, mask
+
+
+def data_generator(image_list, mask_list):
+    dataset = tf.data.Dataset.from_tensor_slices((image_list, mask_list))
+    dataset = dataset.map(load_data, num_parallel_calls=tf.data.AUTOTUNE)
+    dataset = dataset.batch(BATCH_SIZE, drop_remainder=True)
+    dataset = dataset.repeat(1)
```
Comment (fchollet) [inline examples/vision/deeplabv3_plus.py:87]: Why?

File: examples/vision/deeplabv3_plus.py [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+along object boundaries.
+
+![](https://github.com/lattice-ai/DeepLabV3-Plus/raw/master/assets/deeplabv3_plus_diagram.png)
+![](https://github.com/lattice-ai/DeepLabV3-Plus/raw/master/assets/deeplabv3_plus_diagram.png)
+
+
+"""
+
+"""
+**Atrous Convolution:** With Atrous convolution, as we go deeper, we can keep the stride
```
Comment (fchollet) [inline examples/vision/deeplabv3_plus.py:112]: It's called "dilated convolution". "Convolution a trous" (literally "sparse convolution") is a misnomer.

File: examples/vision/deeplabv3_plus.py [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+def convolution_block(
+    block_input,
+    n_filters=256,
+    kernel_size=3,
+    dilation_rate=1,
+    padding="same",
+    use_bias=False,
+):
+
+    x = tf.keras.layers.Conv2D(
```
Comment (fchollet) [inline examples/vision/deeplabv3_plus.py:132]: You can import the `layers` module so you save a few characters each time you instantiate a layer

File: examples/vision/deeplabv3_plus.py [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+
+The reason for using **Atrous Spatial Pyramid Pooling** is that it is discovered as the
+sampling rate becomes larger, the number of valid filter weights (i.e., the weights that
+are applied to the valid feature region, instead of padded zeros) becomes smaller.
+"""
+
+
+def convolution_block(
+    block_input,
+    n_filters=256,
```
Comment (fchollet) [inline examples/vision/deeplabv3_plus.py:125]: For consistency with API conventions, use `num_` as the number prefix

File: examples/vision/deeplabv3_plus.py [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+        use_bias=use_bias,
+        kernel_initializer=tf.keras.initializers.HeNormal(),
+    )(block_input)
+    x = tf.keras.layers.BatchNormalization()(x)
+    return tf.nn.relu(x)
+
+
+def AtrousSpatialPyramidPooling(aspp_input):
+
+    dims = tf.keras.backend.int_shape(aspp_input)
```
Comment (fchollet) [inline examples/vision/deeplabv3_plus.py:146]: Just `aspp_input.shape`

File: examples/vision/deeplabv3_plus.py [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+
+    layer = tf.keras.layers.AveragePooling2D(pool_size=(dims[-3], dims[-2]))(aspp_input)
```
Comment (fchollet) [inline examples/vision/deeplabv3_plus.py:148]: That's a tensor, not a layer, name it `x` or something

File: examples/vision/deeplabv3_plus.py [hunk 11]
```diff
@@ -0,0 +1,298 @@
...
+"""
+The encoder features are first bilinearly upsampled by a factor of 4 and then 
+concatenated with the corresponding low-level features from the network backbone that
+have the same spatial resolution. For this example, we would be 
+using Resnet101 pre-trained on ImageNet as the backbone model and we would use 
+the low-level features from the Conv2 block of the backbone.
+"""
+
+
+def DeeplabV3Plus():
```
Comment (fchollet) [inline examples/vision/deeplabv3_plus.py:179]: You probably want to make `image_size` and `num_classes` signature arguments

File: examples/vision/deeplabv3_plus.py [hunk 12]
```diff
@@ -0,0 +1,298 @@
...
+"""
+## Training
+"""
+
+"""
+We would train the model using Sparse Categorical Cross-entropy as the loss function and
+use Adam as the optimizer.
+"""
+
+loss = tf.losses.SparseCategoricalCrossentropy(from_logits=True)
```
Comment (fchollet) [inline examples/vision/deeplabv3_plus.py:219]: `keras.losses. SparseCategoricalCrossentropy`

File: examples/vision/deeplabv3_plus.py [hunk 13]
```diff
@@ -0,0 +1,298 @@
...
+model.compile(
+    optimizer=tf.keras.optimizers.Adam(learning_rate=0.001),
```
Comment (fchollet) [inline examples/vision/deeplabv3_plus.py:221]: Likewise, import `keras` so you can do `keras.optimizers.Adam`

File: examples/vision/deeplabv3_plus.py [hunk 14]
```diff
@@ -0,0 +1,298 @@
...
+
+
+def infer(model, image_tensor):
+    predictions = model.predict(np.expand_dims((image_tensor), axis=0))
+    predictions = np.squeeze(predictions)
+    predictions = np.argmax(predictions, axis=2)
+    return predictions
+
+
+def decode_segmask(mask, colormap, n_classes):
```
Comment (fchollet) [inline examples/vision/deeplabv3_plus.py:257]: decode_segmentation_masks -- in general use fully spelled out names everywhere

File: examples/vision/deeplabv3_plus.py [hunk 15]
```diff
@@ -0,0 +1,298 @@
...
+    layer = tf.keras.layers.UpSampling2D(
+        size=(
+            IMAGE_SIZE // layer.shape[1],
+            IMAGE_SIZE // layer.shape[2]
+        ), interpolation='bilinear')(layer)
+    
+    model_output = tf.keras.layers.Conv2D(
+        NUM_CLASSES, kernel_size=(1, 1), padding='same')(layer)
+    
+    return tf.keras.Model(inputs=model_input, outputs=model_output)
```
Comment (fchollet) [inline examples/vision/deeplabv3_plus.py:208]: `keras.Model`

**Review by fchollet — CHANGES_REQUESTED** [review]
Thanks for the PR! This is a great topic. I found the code quality to be very good -- great work!

File: examples/vision/deeplabv3_plus.py (hunk 1 above)
Comment (soumik12345) [reply examples/vision/deeplabv3_plus.py:10]: Removed break between consecutive text blocks.

File: examples/vision/deeplabv3_plus.py (hunk 2 above)
Comment (soumik12345) [reply examples/vision/deeplabv3_plus.py]: Edited markdown links to be in single line.

File: examples/vision/deeplabv3_plus.py (hunk 3 above)
Comment (soumik12345) [reply examples/vision/deeplabv3_plus.py]: Removed `!`

File: examples/vision/deeplabv3_plus.py (hunk 4 above)
Comment (soumik12345) [reply examples/vision/deeplabv3_plus.py]: Remove unnecessary cast.

File: examples/vision/deeplabv3_plus.py (hunk 5 above)
Comment (soumik12345) [reply examples/vision/deeplabv3_plus.py]: Removed repeat and prefetch

File: examples/vision/deeplabv3_plus.py (hunk 6 above)
Comment (soumik12345) [reply examples/vision/deeplabv3_plus.py]: Changed to dilated convolution.

File: examples/vision/deeplabv3_plus.py (hunk 7 above)
Comment (soumik12345) [reply examples/vision/deeplabv3_plus.py]: Added the following imports at the top and adjusted the layer, model, loss, and optimizer instantiations

```python3
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
```

File: examples/vision/deeplabv3_plus.py (hunk 8 above)
Comment (soumik12345) [reply examples/vision/deeplabv3_plus.py]: Replaced `n_filters` with `num_filters`

File: examples/vision/deeplabv3_plus.py (hunk 9 above)
Comment (soumik12345) [reply examples/vision/deeplabv3_plus.py]: Set `dims = aspp_input.shape`

File: examples/vision/deeplabv3_plus.py (hunk 10 above)
Comment (soumik12345) [reply examples/vision/deeplabv3_plus.py]: Renamed `layer` to `x`.

File: examples/vision/deeplabv3_plus.py (hunk 11 above)
Comment (soumik12345) [reply examples/vision/deeplabv3_plus.py]: Passed `image_size` and `num_classes` as arguments

File: examples/vision/deeplabv3_plus.py (hunk 12 above)
Comment (soumik12345) [reply examples/vision/deeplabv3_plus.py]: Added the following imports at the top and adjusted the layer, model, loss, and optimizer instantiations

```python3
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
```

File: examples/vision/deeplabv3_plus.py (hunk 13 above)
Comment (soumik12345) [reply examples/vision/deeplabv3_plus.py]: Added the following imports at the top and adjusted the layer, model, loss, and optimizer instantiations

```python3
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
```

File: examples/vision/deeplabv3_plus.py (hunk 14 above)
Comment (soumik12345) [reply examples/vision/deeplabv3_plus.py]: Renamed `decode_segmask` to `decode_segmentation_masks`

File: examples/vision/deeplabv3_plus.py (hunk 15 above)
Comment (soumik12345) [reply examples/vision/deeplabv3_plus.py]: Added the following imports at the top and adjusted the layer, model, loss, and optimizer instantiations

```python3
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
```

**Comment by soumik12345** [comment]
Hi @fchollet thanks for the suggestions for improvement. I have updated the example as per your suggestions, can you please review?

File: examples/vision/deeplabv3_plus.py [hunk 16]
```diff
@@ -0,0 +1,290 @@
...
+
+![](https://github.com/lattice-ai/DeepLabV3-Plus/raw/master/assets/deeplabv3_plus_diagram.png)
+![](https://github.com/lattice-ai/DeepLabV3-Plus/raw/master/assets/deeplabv3_plus_diagram.png)
+
+**Dilated Convolution:** With Dilated convolution, as we go deeper, we can keep the
+stride constant but with larger field-of-view without increasing the number of parameters
+or the amount of computation. And finally, we can have larger output feature map which is
+good for semantic segmentation.
+
+The reason for using **Atrous Spatial Pyramid Pooling** is that it is discovered as the
```
Comment (fchollet) [inline examples/vision/deeplabv3_plus.py]: Change "atrous" to "dilated" everywhere.

**Review by fchollet — CHANGES_REQUESTED** [review]
Thanks for the update! I pushed some copyedits. Please pull them first. Note that there are 2 TODOs left to fill in (text explanations).

**Comment by soumik12345** [comment]
Hi @fchollet I have updated the example as per your suggestions:

- Added a brief description of the CIHP dataset.
- Added a brief explanation of the inference and post-processing steps for visualization.
- Changed `AtrousSpatialPyramidPooling` to `DilatedSpatialPyramidPooling`.

Can you please review?

File: examples/vision/deeplabv3_plus.py (hunk 16 above)
Comment (soumik12345) [reply examples/vision/deeplabv3_plus.py]: Changed atrous to dilated everywhere except the references.

**Review by fchollet — COMMENTED** [review]
Thanks for the update! I added a few minor copyedits. Please pull them, and after that you can add the generated files. The example looks great! 👍

**Comment by soumik12345** [comment]
Hi @fchollet , Added the generated files.

**Review by fchollet — APPROVED** [review]
Awesome, I'll take it from here. Thanks for the great contribution!

**Comment by fchollet** [comment]
Upon closer review, I'd like to request one additional change (to be handled in a separate PR): please add a `validation_data` dataset in the call to `fit()`, and when visualizing the results, please use images from a validation or test set, rather than images from the training set. This is necessary to make sure that the model is capable of generalizing.

**Comment by soumik12345** [comment]
> Upon closer review, I'd like to request one additional change (to be handled in a separate PR): please add a `validation_data` dataset in the call to `fit()`, and when visualizing the results, please use images from a validation or test set, rather than images from the training set. This is necessary to make sure that the model is capable of generalizing.

Hi @fchollet , I raised a new PR updating the example as per your suggestions. Please find it at https://github.com/keras-team/keras-io/pull/629
